File: src/gestorAplicacion/servicios/Servicio.java
package gestorAplicacion.servicios;


import gestorAplicacion.libreria.*;
import uiMain.UiMenu;

import java.io.Serializable;
import java.util.ArrayList;

public abstract class Servicio implements Serializable {
    private static ArrayList<EjemplarRevista> ejemplarRevistaDisponibles = new ArrayList<>();
    private static ArrayList<EjemplarLibro> ejemplarLibroDisponibles = new ArrayList<>();
    private Usuario usuario;
    private Titulo tituloEscogido;
    private Ejemplar EjemplarEscogido;
    private Tiquete tiquete;

    //Constructores

    public Servicio(Usuario usuario, Ejemplar ejemplarEscogido,Titulo tituloEscogido) {
        this.usuario = usuario;
        this.EjemplarEscogido = ejemplarEscogido;
        this.tituloEscogido = tituloEscogido;
    }

    //getters y setters
    public static ArrayList<EjemplarRevista> getEjemplarRevistaDisponibles() {
        return ejemplarRevistaDisponibles;
    }

    public static void setEjemplarRevistaDisponibles(ArrayList<EjemplarRevista> ejemplarRevistaDisponibles) {
        Servicio.ejemplarRevistaDisponibles = ejemplarRevistaDisponibles;
    }

    public static ArrayList<EjemplarLibro> getEjemplarLibroDisponibles() {
        return ejemplarLibroDisponibles;
    }

    public static void setEjemplarLibroDisponibles(ArrayList<EjemplarLibro> ejemplarLibroDisponibles) {
        Servicio.ejemplarLibroDisponibles = ejemplarLibroDisponibles;
    }

    public Usuario getUsuario() {
        return usuario;
    }

    public void setUsuario(Usuario usuario) {
        this.usuario = usuario;
    }

    public Titulo getTituloEscogido() {
        return tituloEscogido;
    }

    public void setTituloEscogido(Titulo tituloEscogido) {
        this.tituloEscogido = tituloEscogido;
    }

    public Ejemplar getEjemplarEscogido() {
        return EjemplarEscogido;
    }

    public void setEjemplarEscogido(Ejemplar ejemplarEscogido) {
        EjemplarEscogido = ejemplarEscogido;
    }

    public Tiquete getTiquete() {
        return tiquete;
    }

    public void setTiquete(Tiquete tiquete) {
        this.tiquete = tiquete;
    }

    //metodos
    public static void filtrarLibrosDisponibles(Biblioteca biblioteca){
        ArrayList<EjemplarLibro> ejemplares = new ArrayList<>();
        for(EjemplarLibro ejemplar: UiMenu.getBiblioteca().getEjemplaresLibros()){
            if(!ejemplar.getEstadoEjemplar().isPrestado() && !ejemplar.getEstadoEjemplar().isReservado()){
                ejemplares.add(ejemplar);
            }
        }
        setEjemplarLibroDisponibles(ejemplares);
    }

    public static void filtrarRevistasDisponibles(Biblioteca biblioteca){
        ArrayList<EjemplarRevista> ejemplares = new ArrayList<>();
        for(EjemplarRevista ejemplar: UiMenu.getBiblioteca().getEjemplaresRevistas()){
            if(!ejemplar.getEstadoEjemplar().isPrestado() && !ejemplar.getEstadoEjemplar().isReservado()){
                ejemplares.add(ejemplar);
            }
        }
        setEjemplarRevistaDisponibles(ejemplares);
    }

    //Metodo que implementa la ligadura dinamica
    public String mostrarTituloEscogido(){
        return (this.getTituloEscogido().mostrarse());
    }
}

File: Python/gestorAplicacion/servicios/servicio.py




class Servicio():
    _ejemplarRevistaDisponibles = []
    _ejemplarLibroDisponibles = []

    def __init__(self, usuario, ejemplarEscogido, tituloEscogido, tiquete = None):
        self._usuario = usuario
        self._ejemplarEscogido = ejemplarEscogido
        self._tituloEscogido = tituloEscogido
        self._tiquete = tiquete

    #Getters y setters
    def getUsuario(self):
        return self._usuario

    def getEjemplarEscogido(self):
        return self._ejemplarEscogido

    def getTituloEscogido(self):
        return self._tituloEscogido

    def getTiquete(self):
        return self._tiquete

    @classmethod
    def getEjemplarRevistaDisponibles(cls) -> list:
        return cls._ejemplarRevistaDisponibles

    @classmethod
    def getEjemplarLibroDisponibles(cls) -> list:
        return cls._ejemplarLibroDisponibles

    def setUsuario(self, usuario):
        self._usuario = usuario

    def setEjemplarEscogido(self, ejemplar):
        self._ejemplarEscogido = ejemplar

    def setTituloEscogido(self, titulo):
        self._tituloEscogido = titulo

    def setTiquete(self, tiquete):
        self._tiquete = tiquete

    @classmethod
    def setEjemplarRevistaDisponibles(cls, ejemplarRevistaDisponibles):
        cls._ejemplarRevistaDisponibles = ejemplarRevistaDisponibles

    @classmethod
    def setEjemplarLibroDisponibles(cls, ejemplarLibroDisponibles):
        cls._ejemplarLibroDisponibles = ejemplarLibroDisponibles

    #Metodos
    def filtrarLibrosDisponibles(self, biblioteca):
        ejemplares = []
        for ejemplar in biblioteca.getEjemplaresLibros():
            if (ejemplar.getEstadoEjemplar().isPrestado() == False and ejemplar.getEstadoEjemplar().isReservado() == False):
                ejemplares.append(ejemplar)

        self.setEjemplarLibroDisponibles(ejemplares)

    def filtrarRevistasDisponibles(self, biblioteca):
        ejemplares = []
        for ejemplar in biblioteca.getEjemplaresRevistas():
            if (ejemplar.getEstadoEjemplar().isPrestado() == False and ejemplar.getEstadoEjemplar().isReservado() == False):
                ejemplares.append(ejemplar)

        self.setEjemplarRevistaDisponibles(ejemplares)

    def mostrarTituloEscogido(self):
        return self.getTituloEscogido().mostrarse()

    @classmethod
    def eliminarDeLibrosDisponibles(cls, ejemplarLibro):
        for ejemplarL in cls._ejemplarLibroDisponibles:
            if ejemplarLibro == ejemplarL:
                cls._ejemplarLibroDisponibles.remove(ejemplarLibro)

    @classmethod
    def eliminarDeRevistasDisponibles(cls, ejemplarRevista):
        for ejemplarR in cls._ejemplarRevistaDisponibles:
            if ejemplarRevista == ejemplarR:
                cls._ejemplarRevistaDisponibles.remove(ejemplarRevista)